parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveComparator.java

```diff
@@ -0,0 +1,202 @@
+    public int compare(int i1, int i2) {
+      return Integer.compare(i1, i2);
+    }
+  };
+
+  static final PrimitiveComparator<Integer> UNSIGNED_INT32_COMPARATOR = new IntComparator() {
+    @Override
+    public int compare(int i1, int i2) {
+      // Implemented based on com.google.common.primitives.UnsignedInts.compare(int, int)
+      return Integer.compare(i1 ^ Integer.MIN_VALUE, i2 ^ Integer.MIN_VALUE);
+    }
+  };
+
+  private static abstract class LongComparator extends PrimitiveComparator<Long> {
+    @Override
+    public final int compare(Long o1, Long o2) {
+      return compare(o1.longValue(), o2.longValue());
+    }
+  }
+
+  static final PrimitiveComparator<Long> SIGNED_INT64_COMPARATOR = new LongComparator() {
+    @Override
+    public int compare(long l1, long l2) {
+      return Long.compare(l1, l2);
+    }
+  };
+
+  static final PrimitiveComparator<Long> UNSIGNED_INT64_COMPARATOR = new LongComparator() {
+    @Override
+    public int compare(long l1, long l2) {
+      // Implemented based on com.google.common.primitives.UnsignedLongs.compare(long, long)
+      return Long.compare(l1 ^ Long.MIN_VALUE, l2 ^ Long.MIN_VALUE);
+    }
+  };
+
+  static final PrimitiveComparator<Float> FLOAT_COMPARATOR = new PrimitiveComparator<Float>() {
+    @Override
+    public int compare(Float o1, Float o2) {
+      return compare(o1.floatValue(), o2.floatValue());
+    }
+
+    @Override
+    public int compare(float f1, float f2) {
+      return Float.compare(f1, f2);
+    }
+  };
+
+  static final PrimitiveComparator<Double> DOUBLE_COMPARATOR = new PrimitiveComparator<Double>() {
+    @Override
+    public int compare(Double o1, Double o2) {
+      return compare(o1.doubleValue(), o2.doubleValue());
+    }
+
+    @Override
+    public int compare(double d1, double d2) {
+      return Double.compare(d1, d2);
+    }
+  };
+
+  private static abstract class BinaryComparator extends PrimitiveComparator<Binary> {
+    @Override
+    public final int compare(Binary o1, Binary o2) {
+      return compare(o1.toByteBuffer(), o2.toByteBuffer());
+    }
+
+    abstract int compare(ByteBuffer b1, ByteBuffer b2);
+
+    final int toUnsigned(byte b) {
+      return b & 0xFF;
+    }
+  }
+
+  static final PrimitiveComparator<Binary> LEXICOGRAPHICAL_BINARY_COMPARATOR = new BinaryComparator() {
+    @Override
+    int compare(ByteBuffer b1, ByteBuffer b2) {
+      int l1 = b1.remaining();
+      int l2 = b2.remaining();
+      int p1 = b1.position();
+      int p2 = b2.position();
+      int minL = Math.min(l1, l2);
+
+      for (int i = 0; i < minL; ++i) {
+        int result = unsignedCompare(b1.get(p1 + i), b2.get(p2 + i));
+        if (result != 0) {
+          return result;
+        }
+      }
+
+      return l1 - l2;
+    }
+
+    private int unsignedCompare(byte b1, byte b2) {
+      return toUnsigned(b1) - toUnsigned(b2);
+    }
+  };
+
+  static final PrimitiveComparator<Binary> SIGNED_BINARY_COMPARATOR = new BinaryComparator() {
```

Comment:
I don't think we have a need for this comparator. The regular ByteBuffer comparator implements the signed byte-wise comparison that is required by the spec.

Reply:
Maybe the name is a bit misleading. This comparator compares the values as signed integers (e.g. BigInteger). We need it if we would like to have similar behaviour for `DECIMAL`s on `INT32`/`INT64` and on `FIX_LEN_BYTE_ARRAY`/`BINARY`.
What a regular `ByteBuffer` comparator or the `compareTo` method of `Binary` do is quite useless. They compare byte values one-by-one as signed values which is incorrect either for string like values or numeric values.

Reply:
To second Gabor's opinion, I agree that this is needed. The byte-wise signed comparison is not appropriate, we need a whole-value signed comparison.

Regarding the comment that the ByteBuffer comparator implements *signed* byte-wise comparison as required by the spec: The [spec](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L756) actually requires *unsigned* byte-wise comparison. (These specific lines were last modified by me, but they was present [earlier](https://github.com/apache/parquet-format/commit/bef5438990116725af041cdd8ced2bca0ed2608a#diff-0f9d1b5347959e15259da7ba8f4b6252L40) as well.)

Reply:
Makes sense. If this is a comparator for encoded decimals, integers, or longs, it should have a more specific name.